File: Assigment 2: Deques and Randomized Queues/Java/RandomizedQueue.java
/**
 * This class represents a randomized queue, which allows adding and removing items in random order.
 * It uses an array-based implementation with resizing to efficiently manage the items.
 * The class provides methods for enqueueing, dequeuing, sampling, and iterating over the items.
 *
 * @param <Item> the type of items stored in the queue
 */

import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.StdRandom;

import java.util.Iterator;

public class RandomizedQueue<Item> implements Iterable<Item> {
    private Item[] s; // Array to store items
    private int n = 0; // Number of items in the queue

    /**
     * Constructs an empty randomized queue.
     */
    public RandomizedQueue() {
        s = (Item[]) new Object[1];
    }

    /**
     * Checks if the randomized queue is empty.
     *
     * @return true if the queue is empty, false otherwise
     */
    public boolean isEmpty() {
        return n == 0;
    }

    /**
     * Returns the number of items in the randomized queue.
     *
     * @return the number of items
     */
    public int size() {
        return n;
    }

    /**
     * Adds an item to the randomized queue.
     *
     * @param item the item to be added
     * @throws IllegalArgumentException if the input item is null
     */
    public void enqueue(Item item) {
        if (item == null) throw new IllegalArgumentException();
        if (n == s.length) resize(2 * s.length);
        s[n++] = item;
    }

    // Private helper method to resize the array
    private void resize(int capacity) {
        Item[] copy = (Item[]) new Object[capacity];
        for (int i = 0; i < n; i++) copy[i] = s[i];
        s = copy;
    }

    /**
     * Removes and returns a random item from the queue.
     *
     * @return the removed item
     * @throws java.util.NoSuchElementException if the queue is empty
     */
    public Item dequeue() {
        if (isEmpty()) {
            throw new java.util.NoSuchElementException();
        }
        int index = StdRandom.uniformInt(0, n);
        Item item = s[index];
        s[index] = s[--n];
        s[n] = null;
        if (n > 0 && n == s.length / 4) resize(s.length / 2);
        return item;
    }

    /**
     * Returns a random item from the queue without removing it.
     *
     * @return a random item
     * @throws java.util.NoSuchElementException if the queue is empty
     */
    public Item sample() {
        if (isEmpty()) {
            throw new java.util.NoSuchElementException();
        }
        int index = StdRandom.uniformInt(0, n);
        Item item = s[index];
        return item;
    }

    /**
     * Returns an independent iterator over items in random order.
     *
     * @return an iterator
     */
    public Iterator<Item> iterator() {
        return new ReverseArrayIterator();
    }

    // Private inner class for the iterator
    private class ReverseArrayIterator implements Iterator<Item> {
        private int i = n;
        private Item[] copy;

        public ReverseArrayIterator() {
            copy = (Item[]) new Object[n];
            for (int j = 0; j < n; j++) {
                copy[j] = s[j];
            }
        }

        public boolean hasNext() {
            return i > 0;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }

        public Item next() {
            if (!hasNext()) throw new java.util.NoSuchElementException();
            int index = StdRandom.uniformInt(0, i);
            Item item = copy[index];
            copy[index] = copy[--i];
            copy[i] = item;
            return item;
        }
    }

    // Unit Testing
    public static void main(String[] args) {
        RandomizedQueue<Integer> randomQueue = new RandomizedQueue<>();
        randomQueue.enqueue(5);
        randomQueue.enqueue(4);
        randomQueue.enqueue(3);
        randomQueue.enqueue(2);
        randomQueue.enqueue(1);
        for (int i : randomQueue) {
            StdOut.println("Item: " + i);
        }
        StdOut.println("randomQueue is empty? " + randomQueue.isEmpty());
        StdOut.println("Size = " + randomQueue.size());
        StdOut.println("Show a random element: " + randomQueue.sample());
        StdOut.println("Remove a random element: " + randomQueue.dequeue());
        StdOut.println("Size = " + randomQueue.size());
    }
}

File: Assignment 2: Deques and Randomized Queues/Python/RandomizedQueue.py
"""
The `RandomizedQueue` class represents a queue where the item removed is chosen uniformly at random
from items in the data structure. It uses a dynamic array (Python list) for efficient operations.

Usage:
- Create a randomized queue: `random_queue = RandomizedQueue()`
- Add an item: `random_queue.enqueue(item)`
- Remove a random item: `random_queue.dequeue()`
- Sample a random item without removing: `random_queue.sample()`
- Check if the queue is empty: `random_queue.is_empty()`
- Get the number of items: `random_queue.size()`
- Iterate over items in random order: `for item in random_queue.iterator(): ...`

Note: The `RandomizedQueue` class ensures that each iterator returns the items in uniformly random order.
"""


import random
from copy import deepcopy

class RandomizedQueue:
    def __init__(self):
        """
        Initializes an empty randomized queue.
        """
        self.s = [] # Internal list to store elements
        self.n = 0 # Number of elements in the queue

    def is_empty(self):
        """
        Checks if the randomized queue is empty.

        Returns:
            True if empty, False otherwise.
        """
        return self.n == 0

    def size(self):
        """
        Returns the number of elements in the randomized queue.

        Returns:
            Number of elements.
        """
        return self.n

    def enqueue(self, item):
        """
        Adds an item to the randomized queue.

        Args:
            item: The item to enqueue.

        Raises:
            ValueError: If the item is None.
        """
        if item is None:
            raise ValueError("Cannot enqueue None")

        self.s.append(item)
        self.n += 1

    def dequeue(self):
        """
        Removes and returns a random item from the randomized queue.

        Returns:
            The dequeued item.

        Raises:
            IndexError: If the queue is empty.
        """
        if self.is_empty():
            raise IndexError("Queue is empty")
        index = random.randint(0, self.n - 1)
        item = self.s[index]
        self.s[index] = self.s[self.n - 1]
        self.s.pop()
        self.n -= 1
        return item

    def sample(self):
        """
        Returns a random item from the randomized queue without removing it.

        Returns:
            A randomly sampled item.

        Raises:
            IndexError: If the queue is empty.
        """
        if self.is_empty():
            raise IndexError("Queue is empty")
        index = random.randint(0, self.n - 1)
        return self.s[index]

    def iterator(self):
        """
        Returns an iterator over a shuffled copy of the elements in the randomized queue.
        The original queue remains unchanged.

        Returns:
            An iterator.
        """
        copy = deepcopy(self.s)
        random.shuffle(copy)
        return iter(copy)


# Example usage
if __name__ == "__main__":
    random_queue = RandomizedQueue()
    random_queue.enqueue(5)
    random_queue.enqueue(4)
    random_queue.enqueue(3)
    random_queue.enqueue(2)
    random_queue.enqueue(1)
    for item in random_queue.iterator():
        print(f"Item: {item}")
    print(f"random_queue is empty? {random_queue.is_empty()}")
    print(f"Size = {random_queue.size()}")
    print(f"Show a random element: {random_queue.sample()}")
    print(f"Remove a random element: {random_queue.dequeue()}")
    print(f"Size = {random_queue.size()}")
